protocols/mdns/src/behaviour.rs

```diff
@@ -359,13 +372,20 @@ where
             } else {
                 tracing::info!(%peer, address=%addr, "discovered peer on address");
                 self.discovered_nodes.push((peer, addr.clone(), expiration));
-                discovered.push((peer, addr));
+                discovered.push((peer, addr.clone()));
+
+                self.pending_events
+                    .push_back(ToSwarm::NewExternalAddrOfPeer {
+                        peer_id: peer,
+                        address: addr,
+                    });
             }
         }
 
         if !discovered.is_empty() {
             let event = Event::Discovered(discovered);
-            return Poll::Ready(ToSwarm::GenerateEvent(event));
+            self.pending_events
+                .push_front(ToSwarm::GenerateEvent(event));
```

Comment:
Ah sorry, I just noticed that we don't have a `loop` statement here and that the `continue` that I suggested in https://github.com/libp2p/rust-libp2p/pull/5753#discussion_r1892089202 doesn't make any sense.

Still, we need a way to continue from top after pushing to `pending_events`. 
Right now we just `push_`, and then return `Poll::Pending` below, instead of `Poll::Ready` with the the head of the queue.

There are two alternatives to solve this:
1. wrap the entire function body in a `loop` and add a `continue` statement here, and below after pushing the `Event::Expired`
1. return the event directly with `return Poll::Ready(...)` same as it was prior to this PR. We then still need the `pending_events`, but just for the `ToSwarm::NewExternalAddrOfPeer`

Fine with both, but slight preference for `1.` because it's a pattern that we also use a lot in other behaviors.

Reply:
Also, if we go with `1.`, could you add a brief comment above the line about using `push_front` instead of `push_back`.
Maybe something along the lines of: 

> `/// Push to the front of the queue so that the behavior event is reported before the individual discovered addresses.`

Reply:
I'm aligned that `1.` is a better solution and maintains consistency let me just do that right away!